### PR TITLE
Auth: only connect to HTTPS endpoints when in strict mode

### PR DESCRIPTION
**Pull request by reinkrul**
Other outbound HTTP connections the Nuts node will make are:

- Downloading IRMA schemes; this is secured by the IRMA server
- Downloading CRLs, I couldn't find definitive evidence this should only be done over HTTPS. Under control of the respective CAs, so should be safe according to their CSPs.
- Downloading JSON-LD schemas; under control of the operator (`jsonld.contexts.remoteallowlist`) (good enough?)

To allow this change, I had to make the following changes:
- Move `RequestAccessToken` HTTP invocation (on remote Nuts node) from API to service layer (also leads to cleaner code)
- Now codeclimate (rightfully) complained about `oauth.go` being too large, realized that it implements multiple OAuth2 roles. Split it in `AuthorizationServer` and `RelyingParty`.
- Also: removed unused `GetOAuthEndpointURL `
- Also: moved Auth's (OAuth) API client to `api/client` to circumvent cyclic dependencies

### CODE REVIEW

File: docs/pages/deployment/configuration.rst [hunk 1]
```diff
@@ -85,4 +85,6 @@ As a general safety precaution ``auth.contractvalidators`` ignores the ``dummy``
 requesting an access token from another node on ``/n2n/auth/v1/accesstoken`` does not return any error details,
 ``http.default.cors.origin`` does not allow a wildcard (``*``),
 json-ld context can only be downloaded from trusted domains configured in ``jsonld.contexts.remoteallowlist``,
-and the ``internalratelimiter`` is always on.
\ No newline at end of file
+and the ``internalratelimiter`` is always on.
+
+Interacting with remote Nuts nodes requires HTTPS: it will refuse to connect to plain HTTP endpoints when in strict mode.
```
Comment (gerardsn) [inline docs/pages/deployment/configuration.rst:90]: this is very vague. "internal" endpoints (/internal, /health, /status, ...) do not require https

**Review by gerardsn — COMMENTED** [review]
how does this work with TLS offloading in proxy?

File: docs/pages/deployment/configuration.rst (hunk 1 above)
Comment (gerardsn) [reply docs/pages/deployment/configuration.rst:90]: aah, missed the `remote` part

**Comment by reinkrul** [comment]
> how does this work with TLS offloading in proxy?

I'd about accessing remote endpoints, regardless of how the remote node handles TLS offloading you'll be connecting to an HTTPS endpoint. We don't support outgoing proxies.

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
Using the engine level interface on the Wrapper is not a good idea. This creates spaghetti. I think it would be cleaner to move client logic from the API layer to the services layer where the configuration has been loaded properly.

**Comment by reinkrul** [comment]
I agree, will be a larger refactor though.

File: auth/api/v1/api.go [hunk 2]
```diff
@@ -23,10 +23,10 @@ import (
 	"fmt"
```
Comment (woutslakhorst) [inline auth/api/v1/api.go:23]: sort imports, the imports on this file are a mess

File: auth/api/v1/api.go [hunk 3]
```diff
@@ -58,6 +59,13 @@ const (
 type Wrapper struct {
 	Auth               auth.AuthenticationServices
 	CredentialResolver vcr.Resolver
+	strictMode         bool
```
Comment (woutslakhorst) [inline auth/api/v1/api.go:62]: unused

File: auth/api/v1/api.go [hunk 4]
```diff
@@ -58,6 +59,13 @@ const (
...
+}
+
+// Configure is called by the system to configure the API wrapper.
+func (w *Wrapper) Configure(config core.ServerConfig) error {
```
Comment (woutslakhorst) [inline auth/api/v1/api.go:66]: Can be removed now?

File: core/engine.go [hunk 5]
```diff
@@ -129,15 +129,29 @@ func (system *System) Configure() error {
...
 			err = m.Configure(*system.Config)
 			coreLogger.Debugf("Configured %s", engineName(engine))
 		}
 		return err
 	})
+	if err != nil {
+		return nil
+	}
+	// Configure API routers
+	for _, router := range system.Routers {
```
Comment (woutslakhorst) [inline core/engine.go:145]: May be removed now?

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

File: core/engine.go (hunk 5 above)
Comment (reinkrul) [reply core/engine.go]: Forgot to revert, good one

File: auth/services/oauth/relying_party.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+		httpClient.Transport = &http.Transport{
+			TLSClientConfig: s.httpClientTLS,
+		}
+	}
+	authClient, err := client.NewHTTPClient("", s.httpClientTimeout, client.WithHTTPClient(httpClient), client.WithRequestEditorFn(core.UserAgentRequestEditor))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create HTTP client: %w", err)
+	}
+	authServerEndpoint, err := url.Parse(authorizationServerEndpoint)
+	if err != nil {
```
Comment (stevenvegt) [inline auth/services/oauth/relying_party.go:88]: better to check the scheme after the parsing?

**Review by stevenvegt — CHANGES_REQUESTED** [review]
Move the scheme check after the url parsing, so you don't have to perform these string operations

File: auth/api/v1/client/types.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2021 Nuts community
```
Comment (woutslakhorst) [inline auth/api/v1/client/types.go:2]: Yes, we won 2 years

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**